relay/arch-doc/rudder-6-reporting-relayd.md

```diff
@@ -0,0 +1,612 @@
+
+This has several advantages:
+
+* It was very easy to start reporting in first releases, because all tooling was already available on Unix systems
+* Performance (except for the postgresql output module in old versions)
+* Well-known by sysadmins (both protocol and services like rsyslog, syslog-ng, etc.)
+
+Problems include:
+
+* TLS set up for syslog is hard, so all reports are sent in plain-text, while they often contain sensible information. This prevents running Rudder over an untrusted network.
+* There is no source authentication, so anyone in the network can generate fake reports
+* We need to install _nxlog_ along with our Windows agent to provide syslog support
+* Syslog problems, especially with UDP have proven very hard to debug
+* When using TCP, it is easy to overflow the syslog infrastructure and affect global system's network stack
+* Configuring the local syslog daemon requires root access
+* Rudder configuration sometimes conflicts with user configuration
+* Reporting uses a totally different mechanism than other communication, which makes it a installation and maintenance burden
+
+#### Requirements
+
+The replacement must:
+
+* Have full encryption (using TLS 1.2+)
+* Use embedded Rudder tools, and do not require root access
+* Run everywhere (embedded hardware, AIX 5.3, RHEL 3, etc.)
+* Allow non-connected modes
+* Be able to report bigger content (full file content, etc.)
+* Be extensible 
+
+#### Considered solutions
+
+We considered various solutions
+
+- message queues: adds operation complexity, can be hard to understand and debug for users
+
+- gRPC: not good for agent communication: async + big files
+
+### From syslog to https (WebDAV)
+
+Inventories are sent to the server using WebDAV.
+We decided to keep it this way and to use
+the same configuration for sending reports.
+
+* Low impact
+* Fallback mode for limited agents
+* Allows using stored files as on-disk queue
+
+The processing program will read files written by apache
+(like we do for inventories).
+
+This will allow keeping compatibility for old agents,
+and avoids having to change server configuration.
+
+### From https PUT to an https API
+
+* connected mode?
+* http/2?
+* serialized messages?
+* client part in Rust?
+
+This may be done in a second time. Requires a storing format, etc.
+
+Should improve performance significantly, and allow richer reports.
+
+## New `relayd` component
+
+### Why Rust?
+
+We considered several languages but finally settled on Rust
+
+- Scala which is already used in the main application. The main issue is that we want relayd to be able to run on light hardware, sometimes embedded, with low resource consumption, contrary to the root server.
+
+The main problem with Rust is that we want be able to push it to the agents due to the absence of AIX in [supported platforms](https://forge.rust-lang.org/platform-support.html).
```

Comment:
this sentence should be either just after the first one ("we considered ... on Rust. Nonetheless, it's a compromise: we won't be able to use Rust on all agent blabla AIX. Other considered languages were: ....")
Or at the end ("even though rust was the less bad, it still has pb, etc)